### PR TITLE
Fix glob statCache to use fs.Stats

### PR DESCRIPTION
**Pull request by styfle**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

If changing an existing definition:
- [x] Provide URL to documentation which provides context for the suggested change: [`statCache` option](https://github.com/isaacs/node-glob/blob/master/README.md#options)

### CODE REVIEW

File: types/glob/index.d.ts [hunk 1]
```diff
@@ -4,9 +4,13 @@
 //                 voy <https://github.com/voy>
 //                 Klaus Meinhardt <https://github.com/ajafff>
 // Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped
+// TypeScript Version: 2.9
```
Comment (ajafff) [inline types/glob/index.d.ts]: why is this change necessary?

File: types/glob/index.d.ts [hunk 2]
```diff
@@ -34,7 +38,7 @@ declare namespace G {
         silent?: boolean;
         strict?: boolean;
         cache?: { [path: string]: boolean | 'DIR' | 'FILE' | ReadonlyArray<string> };
-        statCache?: { [path: string]: false | { isDirectory(): boolean} | undefined };
+        statCache?: { [path: string]: false | Stats | undefined };
```
Comment (ajafff) [inline types/glob/index.d.ts]: I intentionally changed it to this type. I use this option as input to mock certain files. Since the implementation only uses `isDirectory` I don't need to provide a complete implementation of `Stats`.

I guess your use case is the exact opposite and you want to reuse the statCache in your app to avoid calling `fs.stat` again?

**Review by ajafff — CHANGES_REQUESTED** [review]
This is kind of a breaking change (at least for me).
It would help to know your use case and reasoning for this change.

File: types/glob/index.d.ts (hunk 2 above)
Comment (styfle) [reply types/glob/index.d.ts]: > I guess your use case is the exact opposite and you want to reuse the statCache in your app to avoid calling fs.stat again?

Correct 👍

File: types/glob/index.d.ts (hunk 1 above)
Comment (styfle) [reply types/glob/index.d.ts]: I'm not sure. I believe this happened when importing Stat 🤔 

Take a look at the previous CI: https://travis-ci.org/DefinitelyTyped/DefinitelyTyped/builds/508680421

Comment (styfle) [reply types/glob/index.d.ts]: @ajafff I just changed the import so it works with all TS Versions again 👍

**Comment by styfle** [comment]
> It would help to know your use case and reasoning for this change.

My use case is almost identical to the test I added.

I am reading from the `statCache` so this PR is fixing the type declaration to be accurate by adding the missing methods/properties such as `isFile()`, `mode`, etc.

**Comment by styfle** [comment]
@vvakame Any thoughts on this? As far as I can tell, this is the correct type.

**Comment by weswigham** [comment]
@styfle could you not 
```ts
	const pattern = "test/**";
    const options: glob.IOptions & { statCache: {[idx: string]: fs.Stats} } = { statCache: {}, stat: true, dot: true };
	console.log(pattern);

	const mg = new Glob(pattern, options, (er, matches) => {
		if (er) {
			console.error(er);
			return;
		}
        const file1 = matches[0];
        const stat = options.statCache[file1];
        if (stat && stat.isFile()) {
            console.log("is file");
        } else if (stat && stat.isDirectory()) {
            console.log("is directory");
        }
	});
	console.log("after");
```
? That should work just fine, and the library _really_ doesn't need a full stats object provided to it.

**Comment by styfle** [comment]
Hi @weswigham 

I am wrapping another function that does the following:

```ts
export function globHelper(opts: string | glob.IOptions) {
  let options: glob.IOptions;
  if (typeof opts === 'string') {
    options = { cwd: opts };
  } else {
    options = opts;
  }

  options.symlinks = {};
  options.statCache = {};
  options.stat = true;
  options.dot = true;
  // ...
}
```

Sure it doesn't _need_ to have anything. I can use `as`, `any`, even `!` to force it to work the way that I need.

In fact, that is exactly what is happening now as you can see here: https://github.com/zeit/now-builders/blob/a56b32f670627d29a700675ed232cc66a335944e/packages/now-build-utils/src/fs/glob.ts#L49

This file was upgraded to typescript from plain javascript a year or so ago.

The reason why I submitted this PR last year was for correctness and for others who run into the same problem when using the `glob` package, one of the [most depended on packages](https://www.npmjs.com/browse/depended) in the world.

**Comment by sandersn** [comment]
@vvakame @voy @ajafff @weswigham can you review this again? I didn't see an explicit signoff from anybody, which a critical package like `glob` needs.

**Comment by weswigham** [comment]
I didn't wanna signoff on the change since, strictly speaking, you don't need the change to accomplish what the OP wants (you can use a more specific options type locally) and the additional data isn't required by the library.

**Review by sheetalkamat — CHANGES_REQUESTED** [review]
Its not clear if this change is needed.

**Comment by styfle** [comment]
Hi @sheetalkamat 

What changes are you requesting?

**Comment by sheetalkamat** [comment]
Needs signoff from @vvakame @voy @ajafff

**Comment by styfle** [comment]
It sounds like no one wants this fix so I'll just close it and use `any` 🤷‍♂️
